src/sources/dnstap/parser.rs

```diff
@@ -0,0 +1,1105 @@
+    .collect();
+    static ref DNSTAP_MESSAGE_TYPE_ID_MAP: HashMap<i32, DnstapMessageType> = vec![
+        (1, DnstapMessageType::AuthQuery),
+        (2, DnstapMessageType::AuthResponse),
+        (3, DnstapMessageType::ResolverQuery),
+        (4, DnstapMessageType::ResolverResponse),
+        (5, DnstapMessageType::ClientQuery),
+        (6, DnstapMessageType::ClientResponse),
+        (7, DnstapMessageType::ForwarderQuery),
+        (8, DnstapMessageType::ForwarderResponse),
+        (9, DnstapMessageType::StubQuery),
+        (10, DnstapMessageType::StubResponse),
+        (11, DnstapMessageType::ToolQuery),
+        (12, DnstapMessageType::ToolResponse),
+        (13, DnstapMessageType::UpdateQuery),
+        (14, DnstapMessageType::UpdateResponse),
+    ]
+    .into_iter()
+    .map(|(k, v)| (k, v))
+    .collect();
+}
+
+pub struct DnstapParser<'a> {
+    event_schema: &'a DnstapEventSchema,
+    parent_key_path: Vec<PathComponent>,
+    log_event: &'a mut LogEvent,
+}
+
+impl<'a> DnstapParser<'a> {
+    pub fn new(event_schema: &'a DnstapEventSchema, log_event: &'a mut LogEvent) -> Self {
+        Self {
+            event_schema,
+            parent_key_path: Vec::new(),
+            log_event,
+        }
+    }
+
+    fn insert<V>(&mut self, key: &'static str, value: V) -> Option<Value>
+    where
+        V: Into<Value> + Debug,
+    {
+        let mut node_path = self.parent_key_path.clone();
+        node_path.push(PathComponent::Key(key.to_string()));
+        self.log_event.insert_path(node_path, value)
+    }
+
+    pub fn parse_dnstap_data(&mut self, frame: Bytes) -> Result<()> {
+        //parse frame with dnstap protobuf
+        let proto_msg = Dnstap::decode(frame.clone())?;
+
+        if let Some(server_id) = proto_msg.identity {
+            self.insert(
+                self.event_schema
+                    .dnstap_root_data_schema()
+                    .server_identity(),
+                String::from_utf8(server_id).unwrap_or_default(),
+            );
+        }
+
+        if let Some(version) = proto_msg.version {
+            self.insert(
+                self.event_schema.dnstap_root_data_schema().server_version(),
+                String::from_utf8(version).unwrap_or_default(),
+            );
+        }
+
+        if let Some(extra) = proto_msg.extra {
+            self.insert(
+                self.event_schema.dnstap_root_data_schema().extra(),
+                String::from_utf8(extra).unwrap_or_default(),
+            );
+        }
+
+        let dnstap_data_type_id: i32 = proto_msg.r#type;
+        let mut need_raw_data = false;
+        self.insert(
+            self.event_schema.dnstap_root_data_schema().data_type_id(),
+            dnstap_data_type_id,
+        );
+
+        if let Some(dnstap_data_type) = to_dnstap_data_type(dnstap_data_type_id) {
+            self.insert(
+                self.event_schema.dnstap_root_data_schema().data_type(),
+                format!("{:?}", dnstap_data_type),
```

Comment:
Why debug format here?

Reply:
Will change the code to not use debug format here.

Reply:
Updated